http/fetch/archive_fetcher.go

```diff
@@ -89,34 +89,70 @@ func (r *ArchiveFetcher) Fetch(archiveURL, checksum, dir string) error {
 		return fmt.Errorf("failed to download archive from %s, status: %s", archiveURL, resp.Status)
 	}
 
-	var buf bytes.Buffer
+	f, err := os.CreateTemp("", "fetch.*.tmp")
+	if err != nil {
+		return fmt.Errorf("failed to create temp file: %w", err)
+	}
+	defer os.Remove(f.Name())
+
+	// Save temporary file, but limit download to the max download size.
+	if r.maxDownloadSize > 0 {
+		// Headers can lie, so instead of trusting resp.ContentLength,
+		// limit the download to the max download size and error in case
+		// there are still bytes left.
+		// Note that discarding of remaining bytes in resp.Body is a
+		// requirement for Go to effectively reuse HTTP connections.
+		_, err = io.Copy(f, io.LimitReader(resp.Body, int64(r.maxDownloadSize)))
+		n, _ := io.Copy(io.Discard, resp.Body)
+		if n > 0 {
+			return fmt.Errorf("artifact is %d bytes greater than the max download size of %d bytes", n, r.maxDownloadSize)
+		}
+	} else {
+		_, err = io.Copy(f, resp.Body)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to copy temp contents: %w", err)
+	}
 
-	// verify checksum matches origin
-	if err := r.verifyChecksum(checksum, &buf, resp.Body); err != nil {
+	// We have just filled the file, to be able to read it from
+	// the start we must go back to its beginning.
+	_, err = f.Seek(0, 0)
+	if err != nil {
+		return fmt.Errorf("failed to seek back to beginning: %w", err)
+	}
+
+	// Ensure that the checksum of the downloaded file matches the
+	// known checksum.
+	if err := r.verifyChecksum(checksum, f); err != nil {
 		return err
 	}
 
-	// extract
-	if err = tar.Untar(&buf, dir, tar.WithMaxUntarSize(r.maxUntarSize)); err != nil {
-		return fmt.Errorf("failed to extract archive, error: %w", err)
+	// Jump back at the beginning of the file stream again.
+	_, err = f.Seek(0, 0)
+	if err != nil {
+		return fmt.Errorf("failed to seek back to beginning again: %w", err)
+	}
+
+	// Extracts the tar file.
+	if err = tar.Untar(f, dir, tar.WithMaxUntarSize(-1)); err != nil {
```

Comment:
Not really keen on exposing the `MaxUntarSize` as that does not feel like Fetch's responsibility. Forcing an unlimited max untar size is also not good, as that cause issues with highly compressible files. It would be good to get some ideas for this.

Reply:
I see nothing wrong with having `MaxUntarSize` as option here, it's the Fetcher responsibility to download and extract the archive.